pkg/internalapis/cnsvolumeinfo/cnsvolumeinfoservice.go

```diff
@@ -0,0 +1,193 @@
+						return
+					}
+					volumeInfoServiceInstance.volumeIDTovCenter.Store(
+						volumeInfoObject.Spec.VolumeID, volumeInfoObject.Spec.VCenterServer)
+					log.Debugf("successfully added VolumeID: %q, vCenterServer: %q mapping in the cache",
+						volumeInfoObject.Spec.VolumeID, volumeInfoObject.Spec.VCenterServer)
+				},
+				DeleteFunc: func(obj interface{}) {
+					log.Debugf("received delete event for cnsvolumeinfo CR!")
+					var volumeInfoObject cnsvolumeinfov1alpha1.CNSVolumeInfo
+					err := runtime.DefaultUnstructuredConverter.FromUnstructured(
+						obj.(*unstructured.Unstructured).Object, &volumeInfoObject)
+					if err != nil {
+						log.Errorf("failed to cast object to volumeInfoObject. err: %v", err)
+						return
+					}
+					volumeInfoServiceInstance.volumeIDTovCenter.Delete(volumeInfoObject.Spec.VolumeID)
+					log.Debugf("successfully deleted VolumeID: %q, vCenterServer: %q mapping from cache",
+						volumeInfoObject.Spec.VolumeID, volumeInfoObject.Spec.VCenterServer)
+				},
+			}
+			informer.Informer().AddEventHandler(handlers)
+			stopCh := make(chan struct{})
+			informer.Informer().Run(stopCh)
+		}()
+		log.Info("volumeInfo service initialized")
+	}
+	return volumeInfoServiceInstance, nil
+}
+
+// GetvCenterForVolumeID return vCenter for the given VolumeID
+func (volumeInfo *volumeInfo) GetvCenterForVolumeID(ctx context.Context, volumeID string) (string, error) {
+	log := logger.GetLogger(ctx)
+	info, found := volumeInfo.volumeIDTovCenter.Load(volumeID)
+	if found {
+		log.Infof("vCenter: %q found from the cache for VolumeID: %q", info.(string), volumeID)
+		return info.(string), nil
+	} else {
+		return "", logger.LogNewErrorf(log, "Could not find vCenter for VolumeID: %q", volumeID)
+	}
+}
+
+// CreateVolumeInfo creates VolumeInfo CR to persist VolumeID to vCenter mapping
+func (volumeInfo *volumeInfo) CreateVolumeInfo(ctx context.Context, volumeID string, vCenter string) error {
+	log := logger.GetLogger(ctx)
+	log.Infof("creating cnsvolumeinfo for volumeID: %q and vCenter: %q mapping", volumeID, vCenter)
+	cnsvolumeinfo := cnsvolumeinfov1alpha1.CNSVolumeInfo{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      volumeID,
+			Namespace: CSINamespace,
+		},
+		Spec: cnsvolumeinfov1alpha1.CNSVolumeInfoSpec{
+			VolumeID:      volumeID,
+			VCenterServer: vCenter,
+		},
+	}
+	err := volumeInfo.k8sClient.Create(ctx, &cnsvolumeinfo)
+	if err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			log.Errorf("failed to create CR for cnsvolumeInfo CR. Error: %v", err)
+			return err
+		}
+		log.Info("cnsvolumeInfo CR already exists for VolumeID: %q", volumeID)
+		return nil
+	}
+	log.Infof("Successfully created CR for volumeID: %q and vCenter: %q mapping", volumeID, vCenter)
+	return nil
+}
+
+// DeleteVolumeInfo deletes VolumeInfo CR for the given VolumeID
+func (volumeInfo *volumeInfo) DeleteVolumeInfo(ctx context.Context, volumeID string) error {
+	log := logger.GetLogger(ctx)
+	deleteVolumeInfoLock.Lock()
```

Comment:
Why do we need a lock just for deletion?

Reply:
let me think if we need a lock for create operation as well.

Reply:
removed deleteVolumeInfoLock